File: NutriSmart/app/src/main/java/com/example/nutrismart/ui/screens/home/HomeViewModel.kt
package com.example.nutrismart.ui.screens.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.nutrismart.data.UserSession
import com.example.nutrismart.data.model.MealType
import com.example.nutrismart.data.model.UpdateUserRequest
import com.example.nutrismart.data.remote.MealDto
import com.example.nutrismart.data.repository.AuthRepository
import com.example.nutrismart.data.repository.MealRepository
import com.example.nutrismart.data.repository.UserRepository
import com.example.nutrismart.data.repository.WaterRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.LocalTime
import javax.inject.Inject
import kotlin.math.roundToInt

@HiltViewModel
class HomeViewModel @Inject constructor(
    private val waterRepository: WaterRepository,
    private val userRepository: UserRepository,
    private val mealRepository: MealRepository,
    private val authRepository: AuthRepository
) : ViewModel() {
    private val _uiState = MutableStateFlow(HomeUiState())
    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()
    private val currentUserId: Long get() = UserSession.currentUserId
    private val token: String get() = UserSession.token ?: ""
    private var updateWeightJob: Job? = null

    init {
        val savedWater = waterRepository.getWaterIntake()
        val wakeUpTimeStr = authRepository.getWakeUpTime()
        val wakeUpTime = try { LocalTime.parse(wakeUpTimeStr) } catch(e: Exception) { LocalTime.of(8,0) }
        _uiState.update { it.copy(waterConsumedMl = savedWater, wakeUpTime = wakeUpTime) }

        if (currentUserId != -1L) {
            fetchTodayData()
        }
    }

    fun fetchTodayData() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            try {
                val localDate = java.time.LocalDate.now().toString()
                
                val planResponse = mealRepository.getDailyPlan(token, currentUserId, localDate)
                val shoppingListResponse = mealRepository.getShoppingList(token, currentUserId)
                val userResponse = userRepository.getUser(token, currentUserId)

                if (userResponse.isSuccessful) {
                    userResponse.body()?.let { user ->
                        _uiState.update { it.copy(
                            weight = user.weight ?: 0.0,
                            isImperial = user.isImperial
                        ) }
                    }
                }

                if (planResponse.isSuccessful) {
                    planResponse.body()?.let { plan ->
                        _uiState.update { it.copy(
                            breakfast = plan.breakfast, 
                            lunch = plan.lunch, 
                            dinner = plan.dinner, 
                            snack = plan.snack
                        ) }
                    }
                }

                if (shoppingListResponse.isSuccessful) {
                    _uiState.update { it.copy(shoppingList = shoppingListResponse.body()) }
                }
            } catch (e: Exception) { e.printStackTrace() }
            finally { _uiState.update { it.copy(isLoading = false) } }
        }
    }

    fun loadAlternatives(mealTypeStr: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSwapping = true, alternatives = emptyList()) }

            try {
                val mealType = MealType.fromString(mealTypeStr) ?: return@launch
                val response = mealRepository.getMealAlternatives(token, currentUserId, mealType.value)

                if (response.isSuccessful) {
                    _uiState.update { it.copy(alternatives = response.body() ?: emptyList()) }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                _uiState.update { it.copy(isSwapping = false) }
            }
        }
    }

    fun swapMeal(mealTypeStr: String, newMealId: Long) {
        viewModelScope.launch {
            try {
                val mealType = MealType.fromString(mealTypeStr) ?: return@launch
                val response = mealRepository.swapMeal(token, currentUserId, mealType.value, newMealId)

                if (response.isSuccessful) {
                    fetchTodayData()
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun addWater() {
        _uiState.update { state ->
            val newWater = state.waterConsumedMl + state.glassSizeMl
            waterRepository.saveWaterIntake(newWater)
            state.copy(waterConsumedMl = newWater)
        }
    }

    fun removeWater() {
        _uiState.update { state ->
            val newWater = (state.waterConsumedMl - state.glassSizeMl).coerceAtLeast(0)
            waterRepository.saveWaterIntake(newWater)
            state.copy(waterConsumedMl = newWater)
        }
    }

    fun toggleMeal(mealId: Long, mealTypeStr: String, isConsumed: Boolean) {
        val mealType = MealType.fromString(mealTypeStr)
        
        _uiState.update { state ->
            when (mealType) {
                MealType.BREAKFAST -> state.copy(breakfast = state.breakfast?.copy(consumed = isConsumed))
                MealType.LUNCH -> state.copy(lunch = state.lunch?.copy(consumed = isConsumed))
                MealType.DINNER -> state.copy(dinner = state.dinner?.copy(consumed = isConsumed))
                MealType.SNACK -> state.copy(snack = state.snack?.copy(consumed = isConsumed))
                else -> state
            }
        }
        viewModelScope.launch {
            try {
                mealRepository.toggleMealConsumed(token, mealId, isConsumed)
            } catch (e: Exception) { e.printStackTrace() }
        }
    }

    fun toggleShoppingListItem(itemId: Long, checked: Boolean) {
        _uiState.update { currentState ->
            val updatedItems = currentState.shoppingList?.items?.map { if (it.id == itemId) it.copy(checked = checked) else it }
            currentState.copy(shoppingList = currentState.shoppingList?.copy(items = updatedItems ?: emptyList()))
        }
        viewModelScope.launch {
            try {
                mealRepository.toggleShoppingItem(token, itemId, checked)
            } catch (e: Exception) { e.printStackTrace() }
        }
    }

    fun updateHealthData(realSteps: Int, burnedKcal: Int) {
        _uiState.update { it.copy(steps = realSteps, burnedCalories = burnedKcal) }
    }

    fun updateWaterGoalBasedOnWeather(temperature: Int) {
        _uiState.update { state ->
            val newGoal = if (temperature >= 30) 3000 else 2000
            state.copy(waterGoalMl = newGoal)
        }
    }

    fun loadBonusSnacks() {
        viewModelScope.launch {
            _uiState.update { it.copy(isSwapping = true, availableBonusSnacks = emptyList()) }
            try {
                val response = mealRepository.getMealAlternatives(token, currentUserId, "SNACK")

                if (response.isSuccessful) {
                    val allSnacks = response.body() ?: emptyList()
                    val maxCaloriesAllowed = _uiState.value.burnedCalories
                    val filtered = allSnacks.filter { it.calories <= maxCaloriesAllowed }

                    _uiState.update { it.copy(availableBonusSnacks = filtered) }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                _uiState.update { it.copy(isSwapping = false) }
            }
        }
    }

    fun selectBonusSnack(meal: MealDto) {
        _uiState.update { it.copy(bonusSnack = meal) }
    }

    fun toggleBonusSnackConsumed(isConsumed: Boolean) {
        _uiState.update { state ->
            state.copy(bonusSnack = state.bonusSnack?.copy(consumed = isConsumed))
        }
    }

    fun adjustWeight(delta: Double) {
        val currentWeight = _uiState.value.weight
        val newWeight = (currentWeight + delta).coerceAtLeast(30.0)

        val roundedWeight = (newWeight * 10.0).roundToInt() / 10.0

        _uiState.update { it.copy(weight = roundedWeight) }

        updateWeightJob?.cancel()

        updateWeightJob = viewModelScope.launch {
            delay(1500)
            saveWeightToBackend(roundedWeight)
        }
    }

    private suspend fun saveWeightToBackend(newWeight: Double) {
        try {
            if (currentUserId == -1L) return
            val requestBody = UpdateUserRequest(weight = newWeight)
            userRepository.patchUser(token, currentUserId, requestBody)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}

File: NutriSmartIOS/ui/screens/home/HomeViewModel.swift
//
//  HomeViewModel.swift
//  NutriSmartIOS
//

import Foundation
import SwiftUI
import Combine

@Observable
@MainActor
class HomeViewModel {
    var uiState = HomeUiState()
    
    private let waterRepository = WaterRepository.shared
    private let userRepository = UserRepository.shared
    private let mealRepository = MealRepository.shared
    private let authRepository = AuthRepository.shared
    
    private let currentUserId: Int
    private var updateWeightTask: Task<Void, Never>? = nil
    
    init() {
        self.currentUserId = SessionManager.shared.fetchUserId()
        
        let savedWater = waterRepository.getWaterIntake()
        let wakeUpTimeStr = authRepository.getWakeUpTime()
        
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"
        let wakeUpTime = formatter.date(from: wakeUpTimeStr) ?? Calendar.current.date(bySettingHour: 8, minute: 0, second: 0, of: Date())!
        
        uiState.waterConsumedMl = savedWater
        uiState.wakeUpTime = wakeUpTime
        
        if currentUserId != -1 {
            fetchTodayData()
        }
    }
    
    func fetchTodayData() {
        uiState.isLoading = true
        
        let userId = currentUserId
        let token = SessionManager.shared.fetchAuthToken() ?? ""
        
        Task {
            // HealthKit
            if await HealthKitManager.shared.requestAuthorization() {
                let steps = await HealthKitManager.shared.getTodaySteps()
                uiState.steps = steps
                // Assume 0.04 kcal per step for simple demo
                uiState.burnedCalories = Int(Double(steps) * 0.04)
            }
            
            do {
                let user = try await userRepository.getUser(token: token, userId: userId)
                uiState.weight = user.weight ?? 0.0
                uiState.isImperial = user.isImperial
            } catch {
                print("Error fetching user: \(error)")
            }
            
            do {
                let plan = try await mealRepository.getTodayPlan(token: token, userId: userId)
                uiState.breakfast = plan.breakfast
                uiState.lunch = plan.lunch
                uiState.dinner = plan.dinner
                uiState.snack = plan.snack
            } catch {
                print("Error fetching meal plan: \(error)")
            }
            
            do {
                let list = try await mealRepository.getShoppingList(token: token, userId: userId)
                uiState.shoppingList = list
            } catch {
                print("Error fetching shopping list: \(error)")
            }
            
            uiState.isLoading = false
        }
    }
    
    func loadAlternatives(type: String) {
        uiState.isSwapping = true
        uiState.alternatives = []
        
        let userId = currentUserId
        let token = SessionManager.shared.fetchAuthToken() ?? ""
        
        Task {
            do {
                let list = try await mealRepository.getMealAlternatives(token: token, userId: userId, type: type.uppercased())
                uiState.alternatives = list
                uiState.isSwapping = false
            } catch {
                print("Error loading alternatives: \(error)")
                uiState.isSwapping = false
            }
        }
    }
    
    func swapMeal(type: String, newMealId: Int) {
        let userId = currentUserId
        let token = SessionManager.shared.fetchAuthToken() ?? ""
        let date = {
            let f = DateFormatter()
            f.dateFormat = "yyyy-MM-dd"
            return f.string(from: Date())
        }()
        
        Task {
            do {
                _ = try await mealRepository.swapMeal(token: token, userId: userId, type: type.uppercased(), mealId: newMealId, date: date)
                fetchTodayData()
            } catch {
                print("Error swapping meal: \(error)")
            }
        }
    }
    
    func loadBonusSnacks() {
        uiState.isSwapping = true
        uiState.availableBonusSnacks = []
        
        let userId = currentUserId
        let token = SessionManager.shared.fetchAuthToken() ?? ""
        
        Task {
            do {
                let allSnacks = try await mealRepository.getMealAlternatives(token: token, userId: userId, type: "SNACK")
                let maxCalories = uiState.burnedCalories
                uiState.availableBonusSnacks = allSnacks.filter { $0.calories <= maxCalories }
                uiState.isSwapping = false
            } catch {
                print("Error loading bonus snacks: \(error)")
                uiState.isSwapping = false
            }
        }
    }
    
    func updateWaterGoalBasedOnWeather(temp: Int) {
        let newGoal = temp >= 30 ? 3000 : 2000
        uiState.waterGoalMl = newGoal
    }
    
    func addWater() {
        uiState.waterConsumedMl += uiState.glassSizeMl
        waterRepository.saveWaterIntake(water: uiState.waterConsumedMl)
    }
    
    func removeWater() {
        uiState.waterConsumedMl = max(0, uiState.waterConsumedMl - uiState.glassSizeMl)
        waterRepository.saveWaterIntake(water: uiState.waterConsumedMl)
    }
    
    func toggleMeal(mealId: Int, type: MealType, consumed: Bool) {
        switch type {
        case .BREAKFAST: uiState.breakfast = uiState.breakfast.map { var m = $0; m.consumed = consumed; return m }
        case .LUNCH: uiState.lunch = uiState.lunch.map { var m = $0; m.consumed = consumed; return m }
        case .DINNER: uiState.dinner = uiState.dinner.map { var m = $0; m.consumed = consumed; return m }
        case .SNACK: uiState.snack = uiState.snack.map { var m = $0; m.consumed = consumed; return m }
        }
        
        let token = SessionManager.shared.fetchAuthToken() ?? ""
        Task {
            try? await mealRepository.toggleMealConsumed(token: token, mealId: mealId, consumed: consumed)
        }
    }
    
    func toggleShoppingListItem(itemId: Int, checked: Bool) {
        if var list = uiState.shoppingList {
            if let index = list.items.firstIndex(where: { $0.id == itemId }) {
                let updatedItem = ShoppingListItemDto(
                    id: list.items[index].id,
                    category: list.items[index].category,
                    name: list.items[index].name,
                    checked: checked
                )
                var newItems = list.items
                newItems[index] = updatedItem
                uiState.shoppingList = ShoppingListDto(id: list.id, userId: list.userId, items: newItems)
            }
        }
        
        let token = SessionManager.shared.fetchAuthToken() ?? ""
        Task {
            try? await mealRepository.toggleShoppingItem(token: token, itemId: itemId, checked: checked)
        }
    }
    
    func adjustWeight(delta: Double) {
        uiState.weight = max(30.0, uiState.weight + delta)
        let roundedWeight = (uiState.weight * 10).rounded() / 10
        uiState.weight = roundedWeight
        
        updateWeightTask?.cancel()
        updateWeightTask = Task {
            try? await Task.sleep(nanoseconds: 1_500 * 1_000_000)
            if Task.isCancelled { return }
            
            let token = SessionManager.shared.fetchAuthToken() ?? ""
            let request = UpdateUserRequest(
                username: nil, weight: roundedWeight, height: nil, targetWeight: nil,
                activityLevel: nil, maxDailyBudget: nil, dietaryPreferences: nil,
                medicalConditions: nil, dislikedFoods: nil, stepGoal: nil, isImperial: nil, currency: nil
            )
            try? await userRepository.patchUser(token: token, userId: currentUserId, request: request)
        }
    }
}
